support/BRFileService.c

```diff
@@ -260,213 +508,259 @@ fileServiceLoad (BRFileService fs,
-
-            // Do something header specific
-            switch (headerVersion) {
-                case HEADER_FORMAT_1:
-                    // compute the checksum
-                    // compare the checksum
-                   break;
-            }
-
-            // Look up the entity handler
-            BRFileServiceEntityHandler *handler = fileServiceEntityTypeLookupHandler(entityType, version);
-            if (NULL == handler) return fileServiceFailedImpl (fs,  buffer, NULL, "missed type handler");
-
-            // Read the entity from buffer and add to results.
-            void *entity = handler->reader (handler->context, fs, buffer, bytesCount);
-            if (NULL == entity) return fileServiceFailedEntity (fs, buffer, NULL, type, "reader");
-
-            // Update results with the newly restored entity
-            BRSetAdd (results, entity);
-
-            // If the read version is not the current version, update
-            if (updateVersion &&
-                (version != entityType->currentVersion ||
-                 headerVersion != currentHeaderFormatVersion))
-                fileServiceSave (fs, type, entity);
+    if (NULL == entityHandlerCurrent) return fileServiceFailedImpl (fs,  0, NULL, NULL, "missed type handler");
+
+    sqlite3_status_code status;
+
+    pthread_mutex_lock (&fs->lock);
+
+    status = sqlite3_bind_text (fs->sdbSelectAllStmt, 1, type, -1, SQLITE_STATIC);
+    if (SQLITE_OK != status)
+        return fileServiceFailedSDB (fs, 1, status);
+
+    uint8_t  dataBytesBuffer[8196];
+    uint8_t *dataBytes = dataBytesBuffer;
+    size_t   dataBytesCount = 8196;
+
+    while (SQLITE_ROW == sqlite3_step(fs->sdbSelectAllStmt)) {
+        const char *hash = (const char *) sqlite3_column_text (fs->sdbSelectAllStmt, 0);
+        const char *data = (const char *) sqlite3_column_text (fs->sdbSelectAllStmt, 1);
+
+        if (NULL == hash || NULL == data)
+            return fileServiceFailedImpl (fs, 1, (dataBytes == dataBytesBuffer ? NULL : dataBytes), NULL,
+                                          "missed query `hash` or `data`");
+
+        assert (64 == strlen (hash));
+        
+        // Ensure `dataBytes` is large enough for hex-decoded `data`
+        size_t dataCount = strlen (data);
+        assert (0 == dataCount % 2);  // Surely 'even'
+        if ((dataCount/2) > dataBytesCount) {
+            if (dataBytes != dataBytesBuffer) free (dataBytes);
+            dataBytesCount = dataCount/2;
+            dataBytes = malloc (dataBytesCount);
         }
 
-    free (buffer);
-    closedir (dir);
+        // Actually decode `data` into `dataBytes`
+        decodeHex (dataBytes, dataCount/2, data, dataCount);
+
+        size_t offset = 0;
+        BRFileServiceVersion version;
+        uint32_t  entityBytesCount;
+        uint8_t  *entityBytes;
+
+        BRFileServiceHeaderFormatVersion headerVersion = dataBytes[offset];
+        offset += 1;
```

Comment:
I don't think we ensure that dataBytes is large enough before accessing it a offset throughout this loop? Maybe I missed, Github UI sucks.

Reply:
Hmmm, code has this `// Assert entityBytesCount remain in dataBytes`...  I'll fix.